### PR TITLE
Interface density

### PR DESCRIPTION
**Pull request by git-f0x**
Since libcosmic seems to already have a Density enum, I moved most of the related things from Settings to here.

### CODE REVIEW

**Comment by git-f0x** [comment]
I'm marking this ready for review, since it doesn't cause any visual changes, and I'm not sure if I'll be able to rebase this if the Iced rebase happens in the meantime.
Not sure why the header bars don't respond to density changes, so it would be nice to fix that before this is merged, if possible (but I haven't yet been able to find the cause).

**Comment by git-f0x** [comment]
The biggest visual issue (that Compact makes stand out more), is that [this](https://github.com/pop-os/libcosmic/blob/f12de010ec2bd76ee46ef2a33642cf22e55ae87e/src/widget/list/column.rs#L43) needs to have a `min_height` of 32 (potentially less for Compact) to match designs, but I'm not sure how to implement that (as mentioned [here](https://github.com/pop-os/libcosmic/pull/598#issuecomment-2329428222)).

**Comment by git-f0x** [comment]
Changing the `interface_density` config file doesn't seem to do anything. Should ThemeBuilder be changed to take interface density into account (possibly instead of spacing, or to set the spacing to the value of interface density)?

**Comment by wash2** [comment]
> Changing the `interface_density` config file doesn't seem to do anything. Should ThemeBuilder be changed to take interface density into account (possibly instead of spacing, or to set the spacing to the value of interface density)?

Ya the builder could have a method that sets spacing based on the provided density value. And the settings app could set the value.

**Comment by git-f0x** [comment]
How would the builder access the `interface_density` config file (so that it works without Settings)? I tried importing CosmicTk to `/cosmic-theme/src/model/theme.rs` but I only managed to get circular dependencies or "not found".

**Comment by git-f0x** [comment]
Settings already sets the spacing of the theme builder to the chosen density, but the config file isn't used for anything (though I made it write the density to the file).
So should I just move Density to cosmic-theme, and leave everything else as is?
Also regarding [this](https://github.com/pop-os/libcosmic/blob/914d989049ea1a8c47f0690cdbbfe3f862cc491d/src/app/mod.rs#L684), how would I access the current density, so that the padding and spacing could match the header bar (e.g. `padding = if Compact { 2 } else { 8 }`)?

**Comment by git-f0x** [comment]
The NavBar side looks slightly bad with only 2 as padding, but not sure if the header bar horizontal padding should be increased (so that this can be increased as well).
screenshot-2024-09-18-18-05-48

Removing the conflicts caused by recent commits in Settings will take a bit of time.

File: cosmic-theme/src/model/density.rs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                space_xxl: 64,
+                space_xxxl: 128,
+            },
+        }
+    }
+}
+
+impl From<Spacing> for Density {
+    fn from(value: Spacing) -> Self {
+        if (value.space_m - 16) < 1 {
```
Comment (wash2) [inline cosmic-theme/src/model/density.rs:57]:
```suggestion
        if value.space_m.saturating_sub(16) == 0 {
```

File: cosmic-theme/src/model/density.rs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            Self::Compact
+        } else if (value.space_m - 24) < 1 {
```
Comment (wash2) [inline cosmic-theme/src/model/density.rs:59]:
```suggestion
        } else if value.space_m.saturating_sub(24) == 0 {
```

File: src/widget/header_bar.rs [hunk 3]
```diff
@@ -361,7 +362,11 @@ impl<'a, Message: Clone + 'static> HeaderBar<'a, Message> {
         macro_rules! icon {
             ($name:expr, $size:expr, $on_press:expr) => {{
                 #[cfg(target_os = "linux")]
-                let icon = { widget::icon::from_name($name).apply(widget::button::icon) };
+                let icon = {
+                    widget::icon::from_name($name)
+                        .apply(widget::button::icon)
+                        .padding(8)
```
Comment (wash2) [inline src/widget/header_bar.rs:368]: Does this padding not depend on density?

File: src/widget/header_bar.rs [hunk 4]
```diff
@@ -372,6 +377,7 @@ impl<'a, Message: Clone + 'static> HeaderBar<'a, Message> {
                     )))
                     .symbolic(true)
                     .apply(widget::button::icon)
+                    .padding(8)
```
Comment (wash2) [inline src/widget/header_bar.rs:380]: Does this padding not depend on density?

**Review by wash2 — CHANGES_REQUESTED** [review]

File: src/widget/header_bar.rs (hunk 3 above)
Comment (git-f0x) [reply src/widget/header_bar.rs:368]: It does, but the default padding from the widget makes the buttons in the header too small when Compact, and cut off when Spacious.
They still perfectly fit inside the header when Compact, and I think match the buttons in server side decorations, which seem to have the same header height as Compact.

**Comment by wash2** [comment]
> Should `interface_density` be removed from CosmicTk, since it isn't used anywhere, and the config file maybe moved to the builder (possibly also `header_size`)? Not sure what the `interface_density` config file can be used for, since it shouldn't be used by the builder to automatically set spacing so as to not override custom spacing values set by users, and the spacing can be set without it in Settings (`self.theme_builder.spacing = self.density.into()`).

I'm not sure what you mean. https://pop-os.github.io/libcosmic/cosmic/config/fn.interface_density.html returns the CosmicTk interface density value, and is used in your changes. And https://pop-os.github.io/libcosmic/cosmic/config/fn.header_size.html is used too.

**Comment by git-f0x** [comment]
Oh, right, the config file is used in one place. Sorry for the needless question.
Those functions could maybe be moved to the builder (or elsewhere in cosmic-theme), since density is also in cosmic-theme. But probably not necessary.

**Comment by git-f0x** [comment]
Sorry for bothering you, but could this potentially be merged soon, so that apps can update their libcosmic version for easier testing of what needs tweaks (and so that the Settings PR doesn't pull from my fork)?
The one thing I'm not entirely sure about is the way of getting the spacing values. I've seen something like this used in some places:
```
let guard = crate::theme::THEME.lock().unwrap();
let theme = guard.cosmic();
```
and in others something like `theme::active()::cosmic()`. Not sure what it optimal.

**Comment by mmstick** [comment]
We do need some functions to make `cosmic::theme::active().cosmic().spacing` more ergonomic.

**Review by wash2 — DISMISSED** [review]

**Comment by git-f0x** [comment]
Oh, I didn't know that adding a commit automatically dismisses a review. This just fixed the icons in text input becoming cut off or way to small with density changes.

**Review by wash2 — APPROVED** [review]

**Comment by wash2** [comment]
Could you fix the conflicts?

**Comment by git-f0x** [comment]
It says on my end that it has no conflicts. So not sure what to do.

**Review by wash2 — APPROVED** [review]

**Pull request merged**